### ISSUE TITLE
Autosave can still be very slow in 0.4.8

### ISSUE THREAD

**Issue by dg1727** [freeorion/freeorion#2593]
### Environment
* **FreeOrion Version**:  0.4.8 from snapcraft.io (stable release channel) 
* **Operating System**:  Linux Mint based on Ubuntu 16.04 

### Description

As in forum *thread* https://www.freeorion.org/forum/viewtopic.php?f=25&t=11023
And forum *message* https://www.freeorion.org/forum/viewtopic.php?f=25&t=10414#p88247

Autosaves can still be very slow in FreeOrion 0.4.8, even though I think somewhere on the forum said that this problem was thought to have been fixed in 0.4.7.  

I'm playing a single-player game with 150 star systems and 10 AI factions; complexity settings are set to Medium.  I consciously set these to be only a moderate load on the computer.  Max AI aggression level is set to Typical.  

Forum thread 11023 says "you might want to play smaller galaxies, and/or with a lower monster rate", but 150 star systems is at the low end of the suggested galaxy size (15-30 star systems per AI) and from my observation, there are only 1 or 2 small monsters in the galaxy at a time.  

The machine is a 4-core i5 which can play entry-level 3D single-player games pretty well.  

The game is currently at 900+ turns.  The delay on finishing a turn started to be frustratingly slow at about 600 turns.  It now can be 10 minutes per turn, although occasionally it will still be 30 seconds, especially if I have taken a while on the most recent turn.  

Savegame file has been uploaded to http://www.mediafire.com/file/psfgdqm0180euhr/FreeOrion_Living_Player_MyEmpire_0909_2019-Oct-14_00-19-51.428958_%2528punctuation_renamed%2529.sav/file

(Colons in the filename kept MediaFire from accepting the file.  They can also cause complications when doing a data backup to a FAT volume such as a thumb drive.  I know this is a separate issue, but you might consider substituting some other character in autosave filenames.)  

Logs have been uploaded to http://www.mediafire.com/file/m2fkgwfk58kqif2/freeorion_logs.zip/file

The filesystem timestamp on the savegame file is about 45 minutes after the timestamp in the filename.  

I have tried several settings of the first four tickboxes on the Options dialog > Autosave tab, (as well as, on the Other tab, savegame formats of binary vs. uncompressed XML vs. compressed XML), restarting FO between settings changes.  I even tried setting all five tickboxes on the Autosave tab to disabled (the fifth is "autosave on exit").  Even with autosave totally deselected (to confirm:  with settings this way, autosave isn't mentioned in the Messages window), the automated phase of each turn is very time-consuming.  When autosave is totally deselected, the time-consuming bit happens when the Messages window is at "Waiting for other players."  

I've also set the 3 "threads" numbers on the Options > Other tab to 1 and left them there during much of my testing, since the tooltips say that setting the Threads above 1 may cause unpredictable crashes.  (But the defaults were 8, 8, and 2, which contradicts the tooltips.  If crashes are no longer happening often with many threads, you might reword the tooltips.)  

My desktop environment shows a CPU-load graph whenever I haven't got a full-screen app running.  I have mostly been running FO full-screen, but anyway, sometimes when the interactive phase of the turn starts, all 4 cores will stay at 100% usage due to FO.  Also, if I click the Exit button in FO and the FO window disappears, often 1-4 FO processes will still be running at 100% core usage per process.  (When these processes are running after the FO window is gone, I can't start a new instance of FO because FO says an FO server is already running, so I have been killing these processes.  Now that I think about it, I wonder if killing these processes could have corrupted some savegames, but the savegames in the "auto" folder are all about the same size, so hopefully somehow the files haven't been corrupted.)  

Sometimes, when I have FO set to autosave only at the beginning or end of the turn, but not both, the Messages window shows each turn number in 2 consecutive autosaves, but only one of the saves is executed during each turn!  For example:  
* In turn 906, turn 906 is autosaved (That's what the Messages window says.  I haven't compared against the savegame files themselves.)  
* In turn 907, turn 906 is autosaved.  
* In turn 908, turn 907 is autosaved.  
* In turn 909, turn 907 is autosaved.  

As in forum thread 11023, the in-game popup 5-button menu's Save Game button is grayed out starting with the 2nd turn after I load or continue a saved game.  (BTW the tooltip for the grayed-out Save Game button mentions a "green triangle" but I have only seen a green tickmark or an ellipsis in the Empires window.  Maybe the tooltip text could be reworded.)  

Unlike in forum thread 10414, my savegames don't seem to be corrupted.  

At the top-left of the main screen is a toggle button (2 straight-line arrows vs. a circular arrow).  From the description in FO, I'm not clear on what this does, but I have always left it on 2 straight-line arrows.  

I know the game has already been through a lot of design, testing, & refactoring, but perhaps it would make sense if when the game is in single-player mode, games are autosaved synchronously?  I say this because the following items suggest that autosaves are currently asynchronous to gameplay:  
* The Save Game button being grayed out, and the tooltip saying that a savegame is in progress, during the *interactive* segment of the turn, when the user isn't otherwise told that the game is being saved.  
* The usage of many CPU cores at a time, which do all eventually drop to near zero usage if I wait long enough - I assume this CPU usage is for autosave.  
* The fact that this CPU usage continues after the FO window is gone.  

I don't know if asynchronous gamesave in single-player is introducing more problems than it is worth.  But the main thing is to make FO more responsive - how you do it is up to you.

I haven't been checking this e-mail account often in recent times, so I might not pay a lot of attention to your follow-ups.  (I would feel bad about spending a disproportionate amount of my debugging time on game software.  :-)  I will try to check this e-mail over the next few days.  

Thanks for looking into this.

**Comment by Vezzra** [comment]
That's a very tricky issue, as a typical FO game, especially with the game setting you cited (150 systems, 10 AI) only lasts 200/250 turns. Having a game run for 600 or even 900 turns is an extremely rare edge case, meaning practically no one ever runs into that issue...

I don't know if any optimizations have been done since 0.4.8 that might have alleviated these performance issues, so it would be important to verify if they still exist in current master. If you can, please try again with recent test builds.

**Comment by geoffthemedio** [comment]
Yes; test results for a recent test build or equivalent commit would be much more helpful...

**Comment by agrrr3** [comment]
@dg1727 You can simply switch to the snapcraft beta channel to have a 0.4.9 test release. Note though that gets automatically updated each week. It would be great if you could try out playing your setup on that.

**Comment by geoffthemedio** [comment]
I haven't looked at the logs yet, but from the description, it appears that a bug part of the system resource usage and turn processing delays are due to the AI players doing the turns. If this is taking a long time, the AIs may also not be responding promptly to the server asking them for save state data, leading to long apparent save times. I think some aspects of how save info requests are timed may have changed since v0.4.8, however...

@o01eg could perhaps comment

**Comment by Oberlus** [comment]
Regardless of how many users run often into the reported issue (tens of minutes of waiting times between turns), the issue is still there to a lesser extend (having to wait minutes) towards the end of most normal games. So, if it is possible, alleviating this issue would benefit every player of FO.

I've tried loading the provided save game with the latest weekly test version but it isn't compatible. I've started a game with the aim to take it to 1000+ turn and see.

For sure, either the save or the AI or both processing times increase (looks like supralinearly) with the turn number, regardless of galaxy size. Same for save file size. It does look as if the file size grows faster than the number of objects (ships, colonies, monsters, etc.) in the galaxy.

I'm thinking of possible causes of these huge delays/file sizes. Does the system keep record of every created object even if it is out of game/destroyed? I'm thinking mosty of fighters and ships/monsters carcasses that could remain in the objects containers increasing I/O operations.

**Comment by o01eg** [comment]
@geoffthemedio Yes, in 0.4.8 server waits AI to response save data. In 0.4.9 save data sends with turn order together.

**Comment by geoffthemedio** [comment]
@Oberlus Most objects' state, excluding fighters, is remembered in the actual gamestate universe and by each empire after they are destroyed.

**Comment by dg1727** [comment]
Upon closer inspection of the long-running threads, it seems they are the AI computations -- *not* autosave.  This would explain why the delay happens on each turn even when autosave is completely turned off, and why my savegames aren't corrupted after I kill some of those threads.  

It looks like the duration of the AI calcs is based on the number of ships & planets held by that faction; *and* it looks like a faction that's set to be more aggressive takes disproportionately longer.  

The gameplay workaround would basically be to destroy the biggest AI faction first.  Making peace with that faction probably wouldn't help, since the AI would still have to calculate, each turn, what to do with the same large number of ships/planets/buildings.  Anyway, attacking the biggest enemy first is opposite to the approach I usually take with this kind of game.  Maybe someday the AI will be more optimised & will calculate faster.  

Until then, a progress bar (maybe right there in the Empires dialog) for each AI faction might be a good enhancement.  

The Empires dialog does update the computing/done status of each faction in real-time; the turn can advance when all factions have a tickbox (meaning "done") listed against them, instead of an ellipsis (meaning "computing").  

It would be nice if the tooltip when the "Save Game" button is greyed out would reflect specifically whether some AIs were still computing their moves for the turn, rather than suggesting "the problem might be either this or that."  

Also, if the user clicks Resign in this circumstance, a message box appears that is like "A savegame is in progress.  Click this button to close without saving," but that's all it says.  The user should also be told that she/he can wait until the autosave is done (here again a progress bar would be a nice extra), at which time the game will automatically exit.  It would be best if there were also a Cancel button (since the user will expect this option) that would return to the game without waiting for the autosave to finish.  

The original title of this issue was about game-saving.  Even though the reason for opening this issue turned out to be about AI instead of game-saving, my first comment mentions at least one problem in game-saving, in which sometimes the turn being saved, as stated in the Messages window, increments only every other turn of gameplay.  

In general, there's a lot to like about FO, such as the smooth, detailed updating of the play screen.  Overall, this game seems to have fewer & less noticeable bugs than other games I've played that are about as mature in the development process.

**Comment by dg1727** [comment]
I'm sorry, my browser didn't auto-update with the comments since my initial one, and I didn't know people had replied to this issue.  I think my latest comments are still relevant, though.  

To be more specific, one AI faction was more aggressive than others & had captured something like half the galaxy, and produced very many ships (as told by the Graphs window).  This faction is the one that spent the longest in computations each turn, as shown by the real-time "..." in the Empires dialogue.  

There was a second faction that was by far the closest to that first one in calc duration, yet was below second-place in number of ships/planets.  But that second one seemed to play aggressively, hence my speculation about aggressive/cautious AI stance affecting the length of the pause each turn.  

Since you say my supplied savegame won't load in a current beta version, I probably couldn't test any differently than what @Oberlus has already started.  

I was going to edit the issue title to something like "AI calcs @ end of each turn can be very slow" but since a discussion has already started amongst the more experienced people here, I'll let you decide whether to change the title.

**Comment by Oberlus** [comment]
I'm running a multiplayer game with 8 AIs and I as observer in a 150 system galaxy. Turns flow nonstop. Autosave is set for every 20 turns.
My CPU has 4 cores. I can see how it reaches 100% only during AIs operation. I shall test a smaller game with 4 AIs (one per core) to be sure of which empires are taking longer times.
However, with the current run I'm sure the number of objects an AI empire has is not one of the main reasons for the AI delays. At some point near turn 200 the empire with least objects (and least everything, it is dying slowly) was the one causing the greatest delay by a huge difference (13 seconds that AI, less than 3 seconds any other AI, including those with 4 times more ships and systems than the slow one). 20 turns later all empires got back to consume around 3 secs per turn, despite the number of objects being similar.

So maybe there are some AI processes that depending on the situation get into lengthy calculations that does not depend on the assets of the processing AI but on something else (the circumstances of that situation). If the functions that the AI jobs call have time accounting it would be good to have a look on which ones take longer times.

**Comment by geoffthemedio** [comment]
Have a look in the AI log files... There is already some profiling info for save state generation. Adding it to generateOrders isn't difficult:
````
--- "a/C:\\Users\\Geoff\\AppData\\Local\\Temp\\TortoiseGit\\FreeOrionAI-41752cb.000.py"
+++ "b/C:\\Users\\Geoff\\Desktop\\FreeOrionSDK9\\FreeOrion\\default\\python\\AI\\FreeOrionAI.py"
@@ -39,6 +39,7 @@ from common.handlers import init_handlers
 from common.listeners import listener
 from character.character_module import Aggression
 from character.character_strings_module import get_trait_name_aggression, possible_capitals
+from freeorion_tools import profile
 
 main_timer = AITimer('timer', write_log=True)
 turn_timer = AITimer('bucket', write_log=True)
@@ -220,6 +221,7 @@ def handleDiplomaticStatusUpdate(status_update):  # pylint: disable=invalid-name
     diplomatic_corp.handle_diplomatic_status_update(status_update)
 
 
+@profile
 @listener
 def generateOrders():  # pylint: disable=invalid-name
     """
````

I don't really understand the resulting output in the logs, but it's there and looks plausibly useful...

**Comment by Oberlus** [comment]
I no longer think AI's processing is a bottleneck. At least not often. I've run two more unnatended games with 4 AIs. First one lost connection to server before turn 300 (no idea why and didn't bother to check out logs, just started a new game). Second one got to turn 500+ with one AI dominating 95% of the map and being lazy to kill off the remaining planets.
I've spotted the following:
- AIs tend to require more time the more ships they have, but that isn't dramatical, probably logarithmic and certainly not near linear (an AI with 10 ships my take just half the time of an AI with 100 ships).
- Looks like AIs tend to require more time when they struggle to survive. I've found AIs with few objects to require more time than AIs with many times more objects. But that only happens sometimes.
- AIs with distributed empires tend to take more time. Again not significative.
- When there is no real struggle because one AI is obliterating the others, turn processing is faster.

At turn 500 the saves took around 25 seconds, the rest of the processing (AIs and server) roughly the same. Autosave was set to 1 every 20 turns, and the processing times are around 25 seconds each non-saving turn and one minute on the saving turns. All this hold true for the previous 400 turns, i.e. no noticeable increase of turn processing for the last 100 turns (probably because of last point above), so it does not seem to be an evergrowing anything that delays turn processing no matter what.

I think those times are rather acceptable and there is no hurry to look in this issue.

My machine is a i5-2500K at 3.3GHz with 8 GB of RAM and a regular HD (7200 rpm).

**Comment by geoffthemedio** [comment]
A point to consider is that AIs and saving interact, in that each AI serializes its state for the save, and the times they take to do that will probably depend on the numbers of ships and fleets they have.

Were you testing v0.4.8 or a newer test version?

**Comment by Oberlus** [comment]
> Were you testing v0.4.8 or a newer test version?

Latest test version, 2019-10-13, for Windows.

In my time "estimations" I'm just watching the "ready" tick of each AI and the turn updates, counting the time between each and simplistically assigning to AIs the time after turn update and before ready ticks appear, and to the server calculation and the file save (if there is autosave this turn) the time after ready ticks and turn update.

**Comment by Vezzra** [comment]
@geoffthemedio, you have already assigned this to the "Gateway to the Void" milestone (which is intended for issues/PRs that have been closed as rejected/invalid/won't fix/works for me/etc.), but you also tagged it with "testing requested".

Hence I'm unsure what you actually want/mean - should this stay open to receive further testing before closing it as works for me/invalid (in which case it should not yet get assigned to "Gateway to the Void"), or do you consider this as a case closed, so to speak (in which case tagging it with "testing requested" and leaving it open seems rather pointless to me)?

**Comment by geoffthemedio** [comment]
Additional testing would be good before closing. Things may still be problematically slow in some cases. But I don't intend to do anything specifically about this now.

**Comment by Vezzra** [comment]
In that case the issue shouldn't be assigned to "Gateway to the Void" yet.
